crypto/bio/bss_file.c

```diff
@@ -235,15 +235,6 @@ static long file_ctrl(BIO *b, int cmd, long num, void *ptr)
                 _setmode(fd, _O_TEXT);
             else
                 _setmode(fd, _O_BINARY);
-            /*
-             * Reports show that ftell() isn't trustable in text mode.
-             * This has been confirmed as a bug in the Universal C RTL, see
-             * https://developercommunity.visualstudio.com/content/problem/425878/fseek-ftell-fail-in-text-mode-for-unix-style-text.html
-             * The suggested work-around from Microsoft engineering is to
-             * turn off buffering until the bug is resolved.
-             */
-            if ((num & BIO_FP_TEXT) != 0)
-                setvbuf((FILE *)ptr, NULL, _IONBF, 0);
```

Comment:
With this change reverted, why is it necessary to bother to open PEM files in binary mode?
Is there really an expectation that specifically PEM files expect to use `ftel()` and `fseek()`?

Reverting this change restores the (is it still) buggy behaviour of stdio on Windows for non-PEM files, breaking the BIO API semantics for applications that expect seek/tell to work correctly.

If we're now always opening PEM files in binary mode, presumably we can leave this code in place, or if the Windows bug has been fixed on all supported Windows systems, then we can remove this code, but then don't need to bother to open PEM files any differently than other files.

Unless I'm missing something, I'm puzzled why we'd want to **both** revert this work-around **and** open PEM files in binary model.

Reply:
We must open PEM files in binary mode with this change because the decoders & OSSL_STORE depend on being able to do ftell()/fseek(). This hack was added in the 3.0 release and presumably if any other direct use of BIOs in 3rd party applications depended on ftell() fseek() these applications would explicitly call setvbuf() already anyway otherwise they wouldn't work with older OpenSSL versions.

However most likely there aren't many applications (any at all?) using ftell()/fseek() on text mode BIOs but there are applications that use just reads or writes. Their performance was degraded by this hack in 3.0 and the hack should have never been added.

IMO we should do both.

Reply:
I see. Did not known our PEM readers relied on fseek/ftell.

But my question still stands, is it correct to remove disabling buffering in text mode? Might there not now be applications that depend (at some performance cost) on being able to seek text-mode files without their own setvbuf work-arounds?
That is, now we've introduced the setvbuf, some may be relying on it...

The work-around dates back to 3.0, and is now just over 4 years old. Removing it is perhaps a compatibility break?

Reply:
My 2 cents. With respect to the trade offs. The performance regression affects all applications that load PEM files on Windows. It's fairly common for TLS clients to point to `ca.pem` containing a concatenated list of CAs, triggering the perf issue.

The [Windows ftell/fseek bug](https://developercommunity.visualstudio.com/content/problem/425878/fseek-ftell-fail-in-text-mode-for-unix-style-text.html) only affects applications that open files in text mode, but the file contains contains Unix style `\n` line endings. If the file contains `\r\n` line endings, which are native to Windows, then the Windows bug is not triggered.

Additionally with this PR, it would "only" be a compatibility break if a Windows application created a file BIO to load a non-PEM related file and that file contained Unix line endings. Seems possible, but fairly obscure to me.